File: openclaw-skills/risk-assessment/scripts/lib/sheets-client.js
'use strict';

const https = require('https');

function request(method, urlStr, headers, jsonBody) {
  return new Promise((resolve, reject) => {
    const url = new URL(urlStr);
    const body = jsonBody !== undefined ? JSON.stringify(jsonBody) : undefined;
    const req = https.request(
      {
        hostname: url.hostname,
        path: url.pathname + url.search,
        method,
        headers: { ...headers, ...(body ? { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) } : {}) },
      },
      (res) => {
        let data = '';
        res.on('data', (c) => (data += c));
        res.on('end', () => resolve({ status: res.statusCode, body: data }));
      }
    );
    req.on('error', reject);
    if (body) req.write(body);
    req.end();
  });
}

/** Đọc 1 tab bằng API key (chỉ đọc được sheet public). Trả về null nếu 403 (caller tự fallback sang token). */
async function getValuesWithApiKey(fileId, tabName, apiKey) {
  const range = encodeURIComponent(tabName);
  const res = await request('GET', `https://sheets.googleapis.com/v4/spreadsheets/${fileId}/values/${range}?key=${apiKey}`, {});
  if (res.status === 200) return JSON.parse(res.body).values || [];
  if (res.status === 403) return null;
  throw new Error(`Sheets API error đọc "${tabName}" (${res.status}): ${res.body}`);
}

async function getValuesWithToken(fileId, tabName, accessToken) {
  const range = encodeURIComponent(tabName);
  const res = await request('GET', `https://sheets.googleapis.com/v4/spreadsheets/${fileId}/values/${range}`, {
    Authorization: `Bearer ${accessToken}`,
  });
  if (res.status !== 200) throw new Error(`Sheets API error đọc "${tabName}" (${res.status}): ${res.body}`);
  return JSON.parse(res.body).values || [];
}

/** Ghi đè 1 range tường minh (KHÔNG dùng values:append — xem lý do trong SKILL.md). */
async function updateValues(fileId, range, values, accessToken) {
  const encRange = encodeURIComponent(range);
  const res = await request(
    'PUT',
    `https://sheets.googleapis.com/v4/spreadsheets/${fileId}/values/${encRange}?valueInputOption=USER_ENTERED`,
    { Authorization: `Bearer ${accessToken}` },
    { values }
  );
  if (res.status !== 200) throw new Error(`Sheets write error "${range}" (${res.status}): ${res.body}`);
  return JSON.parse(res.body);
}

/** Xóa toàn bộ nội dung 1 range (dùng để reset tab về trạng thái trống hoàn toàn). */
async function clearValues(fileId, range, accessToken) {
  const encRange = encodeURIComponent(range);
  const res = await request('POST', `https://sheets.googleapis.com/v4/spreadsheets/${fileId}/values/${encRange}:clear`, {
    Authorization: `Bearer ${accessToken}`,
  });
  if (res.status !== 200) throw new Error(`Sheets clear error "${range}" (${res.status}): ${res.body}`);
  return JSON.parse(res.body);
}

module.exports = { getValuesWithApiKey, getValuesWithToken, updateValues, clearValues };

File: openclaw-skills/risk-assessment/scripts/lib/sheets_client.py
"""Google Sheets API v4 client — chỉ dùng thư viện chuẩn Python (`urllib`)."""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request

BASE_URL = "https://sheets.googleapis.com/v4/spreadsheets"


class SheetsApiError(RuntimeError):
    def __init__(self, status: int, message: str):
        super().__init__(f"Sheets API error {status}: {message}")
        self.status = status
        self.message = message


def _request(method: str, url: str, token: str, body: dict | None = None) -> dict:
    data = json.dumps(body).encode("utf-8") if body is not None else None
    headers = {"Authorization": f"Bearer {token}"}
    if data is not None:
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        payload = e.read().decode("utf-8")
        try:
            message = json.loads(payload).get("error", {}).get("message", payload)
        except json.JSONDecodeError:
            message = payload
        raise SheetsApiError(e.code, message) from e


def get_spreadsheet_meta(file_id: str, token: str, fields: str = "properties.title,sheets.properties") -> dict:
    url = f"{BASE_URL}/{file_id}?fields={urllib.parse.quote(fields)}"
    return _request("GET", url, token)


def get_values(file_id: str, range_a1: str, token: str) -> list[list]:
    enc_range = urllib.parse.quote(range_a1, safe="")
    url = f"{BASE_URL}/{file_id}/values/{enc_range}"
    result = _request("GET", url, token)
    return result.get("values", [])
